### PR TITLE
AddUserSecrets fails in environments without generic env var and OS special folders

### LINKED ISSUE #66219: AddUserSecrets fails in environment for "DOTNET_USER_SECRETS_FALLBACK_DIR"
### Description

Using `.AddUserSecrets(optional: true)` causes a runtime error when deploying outside my local environment.  The error points to "DOTNET_USER_SECRETS_FALLBACK_DIR"

### Reproduction Steps

Add user secrets to a project using to dotNet CLI, with the configuration builder method `.AddUserSecrets(optional: true)`

### Expected behavior

Making user secrets optional should ignore the fact that no secrets directory exists.  It should not require that a production environment, for instance, have that directory created when the intention is for local development.

### Actual behavior

I get the following runtime error:
```
 "errorType": "InvalidOperationException",
    "errorMessage": "Could not determine an appropriate location for storing user secrets. Set the DOTNET_USER_SECRETS_FALLBACK_DIR environment variable to a folder where user secrets should be stored.",
    "stackTrace": [
      "at Microsoft.Extensions.Configuration.UserSecrets.PathHelper.GetSecretsPathFromSecretsId(String userSecretsId)",
      "at Microsoft.Extensions.Configuration.UserSecretsConfigurationExtensions.AddUserSecretsInternal(IConfigurationBuilder configuration, String userSecretsId, Boolean optional, Boolean reloadOnChange)",
      "at Microsoft.Extensions.Configuration.UserSecretsConfigurationExtensions.AddUserSecrets(IConfigurationBuilder configuration, Assembly assembly, Boolean optional, Boolean reloadOnChange)",
      "at Microsoft.Extensions.Configuration.UserSecretsConfigurationExtensions.AddUserSecrets[T](IConfigurationBuilder configuration, Boolean optional)"...
    ]
```

### Regression?

_No response_

### Known Workarounds

_No response_

### Configuration

.NET 6
Windows 10 x64

### Other information

I've looked on GitHub for similar issues that reference `secrets.json.`  This appears to be a little different but possible related.

### PR DESCRIPTION
**Pull request by pedrobsaila**
Fixes #66219

I could not add unit test : I need to remove special folders (%APPDATA%, %USERPROFILE% in windows) from build machine OS to force using user secrets file

### CODE REVIEW

File: src/libraries/Microsoft.Extensions.Configuration.UserSecrets/src/PathHelper.cs [hunk 1]
```diff
@@ -24,6 +24,22 @@ public class PathHelper
         /// <param name="userSecretsId">The user secret ID.</param>
         /// <returns>The full path to the secret file.</returns>
         public static string GetSecretsPathFromSecretsId(string userSecretsId)
+        {
+            return InternalGetSecretsPathFromSecretsId(userSecretsId, true);
```
Comment (eerhardt) [inline src/libraries/Microsoft.Extensions.Configuration.UserSecrets/src/PathHelper.cs:28]:
```suggestion
            return InternalGetSecretsPathFromSecretsId(userSecretsId, throwIfNoRoot: true);
```

File: src/libraries/Microsoft.Extensions.Configuration.UserSecrets/src/UserSecretsConfigurationExtensions.cs [hunk 2]
```diff
@@ -163,11 +163,16 @@ public static IConfigurationBuilder AddUserSecrets(this IConfigurationBuilder co
 
         private static IConfigurationBuilder AddUserSecretsInternal(IConfigurationBuilder configuration!!, string userSecretsId!!, bool optional, bool reloadOnChange)
         {
-            return AddSecretsFile(configuration, PathHelper.GetSecretsPathFromSecretsId(userSecretsId), optional, reloadOnChange);
+            return AddSecretsFile(configuration, PathHelper.InternalGetSecretsPathFromSecretsId(userSecretsId, !optional), optional, reloadOnChange);
```
Comment (eerhardt) [inline src/libraries/Microsoft.Extensions.Configuration.UserSecrets/src/UserSecretsConfigurationExtensions.cs:166]:
```suggestion
            return AddSecretsFile(configuration, PathHelper.InternalGetSecretsPathFromSecretsId(userSecretsId, throwIfNoRoot: !optional), optional, reloadOnChange);
```

**Review by eerhardt — COMMENTED** [review]
Can you try to brainstorm on how to test this? We have the `RemoteExecutor`, which can start a new process with environment variables set. Can we use that somehow?

**Comment by pedrobsaila** [comment]
> Can you try to brainstorm on how to test this? We have the RemoteExecutor, which can start a new process with environment variables set. Can we use that somehow?

Setting env variables is not sufficient. I need to tweak the OS that executes the test so It does not have a UserProfile/ApplicationData folder. To make this happen, I need to know-how is implemented [SHGetKnownFolderPath](https://docs.microsoft.com/en-us/windows/win32/api/shlobj_core/nf-shlobj_core-shgetknownfolderpath) in windows and getpwuid_r in linux so we can delete the UserProfile/ApplicationData entries from the store that contains them. This can alter the results of other parallel tests that depend on these folders even if I execute my test in a new process

**Comment by pedrobsaila** [comment]
In windows, there is [SHSetKnownFolderPath](https://docs.microsoft.com/en-us/windows/win32/api/shlobj_core/nf-shlobj_core-shsetknownfolderpath?source=docs) which updates UserProfile/ApplicationData folder paths however it does not accept zero-length paths.
In Linux, the libc C library (pwd.h) that provides UserProfile/ApplicationData folder paths (through getpwuid_r) has no method for updating

**Comment by danmoseley** [comment]
I think it is just not feasible to test that.

**Review by maryamariyan — APPROVED** [review]

**Pull request merged**